WordPress/gutenberg
# Automate the cherry-picking process for WordPress and Gutenberg releases

## What problem does this PR solve

Cherry-picking Pull Requests for Gutenberg and WordPress releases is a repetitive process that goes like this:

1. Find all PRs with the `Backport to WP Beta/RC` label
2. Open each in the browser
3. Copy the merge commit hash
4. Run `git cherry-pick $hash` in the CLI
5. If it succeeded, push, let the author know, remove the label
6. If it failed, cherry-pick everything else and retry
7. If it still fails, figure out how to resolve the conflict
8. Back to point 3 until there are no more PRs

This pull request explores a script that reduces the above process to:

1. Run the script
2. Manually resolve conflicts, if any

I successfully used it three times to cherry-pick PRs for WordPress 6.0 RC releases.

## How does the script work?

Here's what `npm run cherry-pick` does:

* Confirms with the developer the current branch aligns with the expectations
* Gets local branches in sync with the remote ones
* Requests the list of PRs to cherry-pick from GitHub API (closed, label=`Backport to WP Beta/RC`)
* Runs `git cherry-pick {commitHash}` for each PR
* It keeps track of the failed cherry-picks and then retries them
* Retrying keeps going as long as at least one cherry-pick succeeds
* Pushes the local branch to `origin`
* (optional) Uses the [`gh` console utility](https://cli.github.com/) to comment on the remote PRs and remove the labels
* Reports the results

Sample output looks as follows:

```
(base) : cloudnik core/plugins/gutenberg wp/6.0; npm run cherry-pick
You are on branch "wp/6.0".

This script will:
• Cherry-pick the merged PRs labeled as "Backport to WP Beta/RC" to this branch
• Push this branch
• Comment on each PR
• Remove the label from each PR

Do you want to proceed? (Y/n)

$ git pull origin wp/6.0 --rebase...
$ git fetch origin trunk...
Found the following PRs to cherry-pick:
   #41198 – Site Editor: Set min-width for styles preview
Fetching commit IDs...
Done!
   #41198 – 860a39665c318d33027d – Site Editor: Set min-width for...
Trying to cherry-pick one by one...
Cherry-picking round 1:
   ✅  cherry-pick commit: afe9b757b4  for PR: #41198 – Site Editor: Set min-width for...
Cherry-picking finished!
Summary:
   ✅  1 PRs got cherry-picked cleanly
   ✅  0 PRs failed

Pushing to origin/wp/6.0
Commenting and removing labels...
✅ 41198: I just cherry-picked this PR to the wp/6.0 branch to get it included in the next release: afe9b757b4
Done!
```

### What tools/permissions are required to run this script?

* [`gh` console utility](https://cli.github.com/) - optional, automates commenting and removing the labels
* Permissions to push to the Gutenberg repository

## Future improvement ideas

* Flexibility: Add an option to exclude specific PRs
* Flexibility: Add an option to specify PR numbers to backport explicitly
* Flexibility: Add an option to specify a custom label or even GH API query
* DevEx: Group reporting functions together, CLI functions together, and GitHub functions together

## Test plan

1. Run the script like `npm run cherry-pick`
2. Confirm it does what this description says
3. Confirm it helps with the release
